// @silevis/reactgrid/dist/lib/Components/HiddenElement.js
import * as React from 'react';
import { isBrowserSafari } from '../Functions/safari';
export var HiddenElement = function (props) {
    var styles = {};
    if (isBrowserSafari() && props.state.viewportElement && props.state.viewportElement.clientHeight !== props.state.cellMatrix.height) {
        var _a = props.state.viewportElement, clientHeight = _a.clientHeight, clientWidth = _a.clientWidth, scrollTop = _a.scrollTop, scrollLeft = _a.scrollLeft;
        var _b = props.state.cellMatrix, cellMatrixHeight = _b.height, cellMatrixWidth = _b.width;
        styles = {
            position: 'absolute',
            height: clientHeight,
            width: clientWidth,
            zIndex: -1,
            top: (scrollTop + clientHeight > cellMatrixHeight) ? cellMatrixHeight - clientHeight : scrollTop,
            left: (scrollLeft + clientWidth > cellMatrixWidth) ? cellMatrixWidth - clientWidth : scrollLeft,
        };
    }
    return React.createElement("input", { className: "rg-hidden-element", style: styles, readOnly: true, ref: props.hiddenElementRefHandler, onBlur: function (e) { return !e.relatedTarget && props.state.hiddenFocusElement.focus(); } });
};
